sources/core.ts

```diff
@@ -23,6 +23,50 @@ export type StateMachine = {
   nodes: Array<Node>;
 };
 
+export type TokenBase = {
+  segmentIndex: number;
+};
+
+export type PathToken = TokenBase & {
+  type: `path`;
+  slice?: undefined;
+};
+
+export type PositionalToken = TokenBase & {
+  type: `positional`;
+  slice?: undefined;
+};
+
+export type OptionDashToken = TokenBase & {
+  type: `dash`;
+  slice?: [number, number];
+  option?: string;
+};
+
+export type OptionToken = TokenBase & {
+  type: `option`;
+  slice?: [number, number];
```

Comment:
I find it a bit counterintuitive that slices are `[start, length]` instead of `[start, end]`.

Both `{Array,String}.prototype.slice` and Rust slices use `[start, end]`.